src/s2python/authorization/client.py

```diff
@@ -1,67 +1,361 @@
-from abc import ABC, abstractmethod
-from typing import Any, Dict
+"""
+S2 protocol client for handling pairing and secure connections.
+"""
 
+import abc
+import json
+import uuid
+import datetime
+from typing import Dict, Optional, Tuple, Union, List, Any
 
-class AbstractConnectionClient(ABC):
-    """Abstract class for handling the /requestConnection endpoint."""
 
-    def request_connection(self) -> Any:
-        """Orchestrate the connection request flow: build → execute → handle."""
-        request_data = self.build_connection_request()
-        response_data = self.execute_connection_request(request_data)
-        return self.handle_connection_response(response_data)
+from jwskate import Jwk
+from pydantic import BaseModel
+
+from s2python.generated.gen_s2_pairing import (
+    ConnectionDetails,
+    ConnectionRequest,
+    PairingRequest,
+    PairingResponse,
+    PairingToken,
+    S2NodeDescription,
+    Protocols,
+)
+
+
+REQTEST_TIMEOUT = 10
+PAIRING_TIMEOUT = datetime.timedelta(minutes=5)
+KEY_ALGORITHM = "RSA-OAEP-256"
+
+
+class PairingDetails(BaseModel):
+    """Contains all details from the pairing process."""
+
+    pairing_response: PairingResponse
+    connection_details: ConnectionDetails
+    decrypted_challenge_str: Optional[str] = None
 
-    @abstractmethod
-    def build_connection_request(self) -> Dict:
-        """
-        Build the payload for the ConnectionRequest schema.
-        Returns a dictionary with keys: s2ClientNodeId, supportedProtocols.
-        """
 
-    @abstractmethod
-    def execute_connection_request(self, request_data: Dict) -> Dict:
+class S2AbstractClient(abc.ABC):
+    """Abstract client for handling S2 protocol pairing and connections.
+
+    Client handles:
+    - HTTP client with TLS
+    - Storage of connection request URI
+    - Storage of public/private key pairs
+    - Challenge solving
+
+    This class serves as an interface that developers can extend to implement
+    S2 protocol functionality with their preferred technology stack.
+    Concrete implementations should override the abstract methods marked
+    with @abc.abstractmethod.
+    """
+
+    # pylint: disable=too-many-instance-attributes
+    # pylint: disable=too-many-arguments
+    def __init__(
```

Comment:
hhm I wonder if we want to couple the JWK and encryption methods inside the abstract class or not.

Reply:
```
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from jwskate import Jwk
```